lib/Optimizer/Dialect/CC/CCOps.cpp

```diff
@@ -1788,6 +1797,72 @@ MutableOperandRange cudaq::cc::ConditionOp::getMutableSuccessorOperands(
   return getResultsMutable();
 }
 
+//===----------------------------------------------------------------------===//
+// OffsetOfOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult cudaq::cc::OffsetOfOp::verify() {
+  Type ty = getInputType();
+  for (std::int32_t i : getConstantIndices()) {
+    if (auto strTy = dyn_cast<cc::StructType>(ty)) {
+      if (i < 0 || static_cast<std::size_t>(i) >= strTy.getMembers().size())
+        return emitOpError("out of bounds for struct");
+      ty = strTy.getMembers()[i];
+      continue;
+    }
+    if (auto arrTy = dyn_cast<cc::ArrayType>(ty)) {
+      if (arrTy.isUnknownSize())
+        return emitOpError("array must have bounds");
+      if (i < 0 || i >= arrTy.getSize())
+        return emitOpError("out of bounds for array");
+      ty = arrTy.getElementType();
+      continue;
+    }
+    if (auto complexTy = dyn_cast<ComplexType>(ty)) {
+      if (i < 0 || i > 1)
+        return emitOpError("out of bounds for complex");
+      ty = complexTy.getElementType();
+      continue;
+    }
+    // Walking off the end of the type.
+    return failure();
+  }
+  return success();
+}
+
+namespace {
+struct FoldTrivialOffsetOf : public OpRewritePattern<cudaq::cc::OffsetOfOp> {
+  using Base = OpRewritePattern<cudaq::cc::OffsetOfOp>;
+  using Base::Base;
+
+  LogicalResult matchAndRewrite(cudaq::cc::OffsetOfOp offOp,
+                                PatternRewriter &rewriter) const override {
+    // If there are no offsets, the offset is 0.
+    if (offOp.getConstantIndices().empty()) {
+      rewriter.replaceOpWithNewOp<arith::ConstantIntOp>(offOp, 0,
+                                                        offOp.getType());
+      return success();
+    }
+
+    // If all indices are 0, then the offset is necessarily 0.
```

Comment:
What if you have a struct of `{int16_t, int32_t}`? The compiler will often insert 16 bits of padding, and depending on endianness, it seems like the padding could be prior to the 1st `int16_t`, and in that case, the offset to the first element could be `2`, no?

Reply:
The endianness of the machine will affect the byte order vs. significance, but not the struct layout. See https://stackoverflow.com/questions/38992147/does-the-endianness-affect-how-structure-members-are-stored-into-the-memory for more discussion.